src/js/hooks/inboundV2/addItems/useInboundAddItemsForm.jsx

```diff
@@ -335,34 +353,74 @@ const useInboundAddItemsForm = ({
     }
   };
 
-  const handleExpirationDateMismatch = (updatedValues, lineItems, hasNonZeroQuantity) => {
-    if (hasNonZeroQuantity) {
-      confirmAction(
-        () => updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems),
-        'react.stockMovement.confirmExpiryDateUpdate.message',
-        'This will update the expiry date across all depots in the system. Are you sure you want to proceed?',
-      );
-    } else {
-      updateInventoryItemsAndTransitionToNextStep(updatedValues, lineItems);
+  /**
+   *  Shows Inventory item expiration date update confirmation modal.
+   *  @param {Array} mismatchedItems - Array of elements with mismatched expiration dates.
+   *  @returns {Promise} - Resolves to true if user confirms the update, false if not.
+   * @public
+   */
+  const confirmExpirationDateSave = (mismatchedItems) => new Promise((resolve) => {
+    setItemsWithMismatchedExpiry(mismatchedItems);
+    setResolveExpirationPromise(() => resolve);
+    setIsExpirationModalOpen(true);
+  });
+
+  /**
+   * Handles the response from the expiration date confirmation modal.
+   * @param {boolean} shouldUpdate - True if the user confirmed the update, false if not.
+   * @public
+   */
+  const handleExpirationModalResponse = (shouldUpdate) => {
+    // Resolve the promise returned by confirmExpirationDateSave.
+    if (resolveExpirationPromise) {
+      resolveExpirationPromise(shouldUpdate);
     }
+    // Close the modal and reset its state.
+    setIsExpirationModalOpen(false);
+    setItemsWithMismatchedExpiry([]);
+    setResolveExpirationPromise(null);
```

Comment:
given that those states always seem to be updated together, I'd consider creating one object based state.